src/main/java/com/feedzai/commons/sql/abstraction/engine/pool/PooledDatabaseEngine.java

```diff
@@ -0,0 +1,436 @@
+ */
+package com.feedzai.commons.sql.abstraction.engine.pool;
+
+import java.sql.Connection;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import javax.annotation.Nullable;
+import org.apache.commons.pool2.impl.GenericObjectPool;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.feedzai.commons.sql.abstraction.FailureListener;
+import com.feedzai.commons.sql.abstraction.batch.AbstractBatch;
+import com.feedzai.commons.sql.abstraction.ddl.DbColumnType;
+import com.feedzai.commons.sql.abstraction.ddl.DbEntity;
+import com.feedzai.commons.sql.abstraction.ddl.DbEntityType;
+import com.feedzai.commons.sql.abstraction.dml.Expression;
+import com.feedzai.commons.sql.abstraction.dml.dialect.Dialect;
+import com.feedzai.commons.sql.abstraction.dml.result.ResultColumn;
+import com.feedzai.commons.sql.abstraction.dml.result.ResultIterator;
+import com.feedzai.commons.sql.abstraction.engine.ConnectionResetException;
+import com.feedzai.commons.sql.abstraction.engine.DatabaseEngine;
+import com.feedzai.commons.sql.abstraction.engine.DatabaseEngineException;
+import com.feedzai.commons.sql.abstraction.engine.DatabaseEngineRuntimeException;
+import com.feedzai.commons.sql.abstraction.engine.DuplicateEngineException;
+import com.feedzai.commons.sql.abstraction.engine.NameAlreadyExistsException;
+import com.feedzai.commons.sql.abstraction.engine.RecoveryException;
+import com.feedzai.commons.sql.abstraction.engine.RetryLimitExceededException;
+import com.feedzai.commons.sql.abstraction.engine.configuration.PdbProperties;
+import com.feedzai.commons.sql.abstraction.engine.handler.ExceptionHandler;
+import com.feedzai.commons.sql.abstraction.entry.EntityEntry;
+import com.feedzai.commons.sql.abstraction.listeners.BatchListener;
+
+/**
+ * Pooled {@link DatabaseEngine} which delegates its operations to a wrapped database engine instance. The
+ * {@link #close()} operation causes the pooled database engine to be returned to pool.
+ *
+ * @author Luiz Silva (luiz.silva@feedzai.com)
+ * @author Francisco Santos (francisco.santos@feedzai.com)
+ * @since 2.8.3
+ */
+class PooledDatabaseEngine implements DatabaseEngine {
+
+    /**
+     * The logger.
+     */
+    private static final Logger LOGGER = LoggerFactory.getLogger(PooledDatabaseEngine.class);
+
+    /**
+     * The pool which owns this pooled database engine.
+     */
+    private final GenericObjectPool<PooledDatabaseEngine> pool;
+
+    /**
+     * The wrapped database engine.
+     */
+    private final DatabaseEngine engine;
+
+    /**
+     * Creates a new {@link PooledDatabaseEngine}.
+     *
+     * @param pool the pool owning the pooled database engine.
+     * @param engine the database engine to be wrapped.
+     */
+    PooledDatabaseEngine(final GenericObjectPool<PooledDatabaseEngine> pool, final DatabaseEngine engine) {
+        this.pool = pool;
+        this.engine = engine;
+    }
+
+    @Override
+    public void close() {
```

Comment:
Since closing the `PooledDatabaseEngine` does not actually close the `DatabaseEngine` object, I would like to understand/see how the database connection actually gets freed when the pooled object containing the `DatabaseEngine` object is evicted or destroyed from the pool.

Reply:
Hi @nahian18 

I did some quick tests and indeed we're not closing the underlying connections after returning the `DatabaseEngine` object to the pool. 😢 

Thanks for point this! Following the [apache pool documentation](https://commons.apache.org/proper/commons-pool/apidocs/org/apache/commons/pool2/PooledObjectFactory.html), I'veoverriddend the `destroyObject` method to close the underlying connection.
After this change my tests show that we aren't leaking connections anymore 😄 

Can you please review the recent pushed changes?

Thanks!

Reply:
Hi again!

Just added more logging for this methods and a unit test to check if the connection is closed once the object is destroyed!

Thanks!